02_bubble/alfredxiao/main_test.go

```diff
@@ -0,0 +1,65 @@
+package main
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBubbleEdgeCases(t *testing.T) {
+	assert.Equal(t, []int{}, bubble([]int{}))
+	assert.Equal(t, []int{1}, bubble([]int{1}))
+	assert.Equal(t, []int{1, 2}, bubble([]int{2, 1}))
+}
+
+func TestBubbleBestCase(t *testing.T) {
+	assert.Equal(t, []int{1, 2, 3, 4, 5, 6, 7, 8, 9}, bubble([]int{1, 2, 3, 4, 5, 6, 7, 8, 9}))
+}
+
+func TestBubbleAverageCase(t *testing.T) {
+	assert.Equal(t, []int{1, 2, 3, 6, 9, 10, 12}, bubble([]int{12, 2, 1, 6, 9, 3, 10}))
+}
+
+func TestBubbleWorstCase(t *testing.T) {
+	assert.Equal(t, []int{1, 2, 3, 4, 5, 6, 7, 8, 9}, bubble([]int{9, 8, 7, 6, 5, 4, 3, 2, 1}))
+}
+
+func TestBubbleDuplicates(t *testing.T) {
+	assert.Equal(t, []int{2, 3, 3, 6, 6, 8, 9}, bubble([]int{3, 2, 3, 6, 8, 6, 9}))
+}
+
+func TestBubbleDoesNotTouchInput(t *testing.T) {
+	ns := []int{3, 2, 1}
+	bubble(ns)
+	assert.Equal(t, []int{3, 2, 1}, ns)
+}
+
+func TestInsertionEdgeCases(t *testing.T) {
+	assert.Equal(t, []int{}, insertion([]int{}))
+	assert.Equal(t, []int{1}, insertion([]int{1}))
+	assert.Equal(t, []int{1, 2}, insertion([]int{2, 1}))
+}
+
+func TestInsertionBestCase(t *testing.T) {
```

Comment:
This test is the same as `TestBubbleBestCase`

I'd suggest passing function as an argument to test: i.e
```
algo := {"bubble": bubble, "insertion": insertion}
for _,f := range algo {
    assert.Equal(t, []int{2, 3, 3, 6, 6, 8, 9}, f([]int{3, 2, 3, 6, 8, 6, 9}))
}
```
as next step you can also keep slices with data and pass to the test:
```
algo := {"bubble": bubble, "insertion": insertion}
slices := [{"input":[]int{3,1,2}, "expected":[]int{1,2,3} }, ..., ..., ... ]
for _,f := range algo {
    for _, s := range slices {
      assert.Equal(t, s.expected, f(s.input))
  }
}
```
as next improvement you can put assert inside subtest
```
algo := {"bubble": bubble, "insertion": insertion}
slices := [{"input":[]int{3,1,2}, "expected":[]int{1,2,3} }, ..., ..., ... ]
for a, f := range algo {
    for _, s := range slices {
      t.Run(a, func(t *testing.T) {
      assert.Equal(t, s.expected, f(s.input))
    }
  }
}
```

Reply:
good suggestion, I have applied some refactoring there

Reply:
I guess you forgot to push. `628a2e8deda6e0152b18ae247b4b6b5e194cdc69` doesn't touch this file.

Reply:
pardon me, pushed now